qbft/round_change.go

```diff
@@ -154,58 +144,84 @@ func hasReceivedProposalJustificationForLeadingRound(
 			roundChangeJustification = append(roundChangeJustification, msg)
 		}
 
-		if isProposalJustificationForLeadingRound(
+		// checked on isValidRoundChange
+		if valueToPropose, err := isProposalJustificationForLeadingRound(
 			state,
 			config,
 			containerRoundChangeMessage,
 			roundChanges,
 			roundChangeJustification,
-			valueToPropose,
+			cdFetcher,
 			valCheck,
 			roundChangeMessage.QBFTMessage.Round,
-		) == nil {
+		); err == nil {
 			// not returning error, no need to
 			return containerRoundChangeMessage, valueToPropose, nil
 		}
 	}
 	return nil, nil, nil
 }
 
-// isProposalJustificationForLeadingRound - returns nil if we have a quorum of round change msgs and highest justified value for leading round
+// isProposalJustificationForLeadingRound - returns proposal value if we have a quorum of round change msgs and
+// highest justified value for leading round
 func isProposalJustificationForLeadingRound(
 	state *State,
 	config IConfig,
 	roundChangeMsg *ProcessingMessage,
 	roundChanges []*ProcessingMessage,
-	roundChangeJustifications []*ProcessingMessage,
-	value []byte,
+	roundChangeJustifications []*ProcessingMessage, cdFetcher *types.DataFetcher,
 	valCheck ProposedValueCheckF,
 	newRound Round,
-) error {
+) ([]byte, error) {
+	if proposer(state, config, roundChangeMsg.QBFTMessage.Round) != state.CommitteeMember.OperatorID {
+		return nil, errors.New("not proposer")
+	}
+
+	valueToPropose, err := chooseValueToPropose(roundChangeMsg.QBFTMessage, cdFetcher, roundChangeMsg.SignedMessage.FullData)
+	if err != nil {
+		return nil, err
+	}
 
 	if err := isReceivedProposalJustification(
 		state,
 		config,
 		roundChanges,
 		roundChangeJustifications,
 		roundChangeMsg.QBFTMessage.Round,
-		value,
+		valueToPropose,
 		valCheck); err != nil {
-		return err
+		return nil, err
 	}
 
 	if proposer(state, config, roundChangeMsg.QBFTMessage.Round) != state.CommitteeMember.OperatorID {
-		return errors.New("not proposer")
+		return nil, errors.New("not proposer")
 	}
 
 	currentRoundProposal := state.ProposalAcceptedForCurrentRound == nil && state.Round == newRound
 	futureRoundProposal := newRound > state.Round
 
 	if !currentRoundProposal && !futureRoundProposal {
-		return errors.New("proposal round mismatch")
+		return nil, errors.New("proposal round mismatch")
 	}
 
-	return nil
+	return valueToPropose, nil
+}
+
+// chooseValueToPropose
+// If justifiedRoundChangeMsg has no prepare justification choose state value
+// If justifiedRoundChangeMsg has prepare justification choose prepared value
+func chooseValueToPropose(roundChangeMsg *Message, cdFetcher *types.DataFetcher, preparedValue []byte) ([]byte, error) {
```

Comment:
since its only used in one place, this shouldn't be a method IMO, or at least the name should reflect that a long-awaited fetch is part of this function.